pkg/nodeagent/registry/containerd_extractor.go

```diff
@@ -133,56 +134,122 @@ func unmountImage(ctx context.Context, snapshotter snapshots.Snapshotter, direct
 }
 
 // CopyFile copies a source file to destination file and sets the given permissions.
-func CopyFile(fs afero.Afero, sourceFile, destinationFile string, permissions os.FileMode) error {
-	sourceFileStat, err := fs.Stat(sourceFile)
+func CopyFile(fs afero.Afero, source, destination string, permissions os.FileMode) error {
+	if destinationFileStat, err := fs.Stat(destination); err == nil {
+		if !destinationFileStat.Mode().IsRegular() {
+			return fmt.Errorf("destination %q exists but is not a regular file", destination)
+		}
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	if err := fs.MkdirAll(path.Dir(destination), permissions); err != nil {
+		return fmt.Errorf("destination directory %q could not be created", path.Dir(destination))
+	}
+
+	tempDir, err := fs.TempDir(nodeagentv1alpha1.TempDir, "copy-image-")
 	if err != nil {
+		return fmt.Errorf("error creating temp directory: %w", err)
+	}
+
+	defer func() { utilruntime.HandleError(fs.Remove(tempDir)) }()
+
+	tmpFilePath := filepath.Join(tempDir, filepath.Base(destination))
+
+	if err := copyFileAndSync(fs, source, tmpFilePath); err != nil {
+		return err
+	}
+
+	if err := fs.Chmod(tmpFilePath, permissions); err != nil {
 		return err
 	}
 
+	return MoveFile(fs, tmpFilePath, destination)
+}
+
+func copyFileAndSync(fs afero.Afero, source, destination string) error {
+	sourceFileStat, err := fs.Stat(source)
+	if err != nil {
+		return err
+	}
 	if !sourceFileStat.Mode().IsRegular() {
-		return fmt.Errorf("source %q is not a regular file", sourceFile)
+		return fmt.Errorf("source %q is not a regular file", source)
 	}
 
-	if destinationFileStat, err := fs.Stat(destinationFile); err == nil {
+	if destinationFileStat, err := fs.Stat(destination); err == nil {
 		if !destinationFileStat.Mode().IsRegular() {
-			return fmt.Errorf("destination %q exists but is not a regular file", destinationFile)
+			return fmt.Errorf("destination %q exists but is not a regular file", destination)
 		}
 	} else if !os.IsNotExist(err) {
 		return err
 	}
 
-	srcFile, err := fs.Open(sourceFile)
+	srcFile, err := fs.Open(source)
 	if err != nil {
 		return err
 	}
-	defer srcFile.Close()
+	defer func() { utilruntime.HandleError(srcFile.Close()) }()
 
-	if err := fs.MkdirAll(path.Dir(destinationFile), permissions); err != nil {
-		return fmt.Errorf("destination directory %q could not be created", path.Dir(destinationFile))
+	if err := fs.MkdirAll(path.Dir(destination), os.ModeDir); err != nil {
+		return fmt.Errorf("destination directory %q could not be created", path.Dir(destination))
 	}
 
-	tempDir, err := fs.TempDir(nodeagentv1alpha1.TempDir, "copy-image-")
+	dstFile, err := fs.OpenFile(destination, os.O_CREATE|os.O_RDWR, 0755)
```

Comment:
When called via `CopyFile`, the destination file already exists. Then the dstFile must be truncated using `os.O_TRUNC` to make sure that no leftovers remain.

Reply:
`CopyFile` first creates a new temporary dir ([ref](https://github.com/gardener/gardener/blob/2c24775e3243897903cec5d3dc94b38da0b8bab9/pkg/nodeagent/registry/containerd_extractor.go#L150)) and copies the file there ([ref](https://github.com/gardener/gardener/blob/2c24775e3243897903cec5d3dc94b38da0b8bab9/pkg/nodeagent/registry/containerd_extractor.go#L157-L159)).
Thus, it should not exist shouldn't it?

Reply:
I'd suggest to use `0700` as filemode, it will be replace by `Chmod` anyways.

Reply:
> Thus, it should not exist shouldn't it?

Oh, indeed I didn't notice the TempDir that's used there. Although in that case `os.O_CREATE|os.O_RDWR|os.O_EXCL` would be much cleaner than trying to check that the destination file does not already exist.

But if I'm not mistaken, this guarantee does not hold for `moveCrossDevice` at the moment? There the destination is not a random location.